components/pd_client/src/tso.rs

```diff
@@ -0,0 +1,236 @@
+use tokio::sync::{mpsc, oneshot, watch};
+use txn_types::TimeStamp;
+
+/// It is an empirical value.
+const MAX_BATCH_SIZE: usize = 64;
+
+const MAX_PENDING_COUNT: usize = 1 << 16;
+
+type TimestampRequest = oneshot::Sender<TimeStamp>;
+
+/// The timestamp oracle (TSO) which provides monotonically increasing timestamps.
+pub struct TimestampOracle {
+    /// The transmitter of a bounded channel which transports requests of getting a single
+    /// timestamp to the TSO working thread. A bounded channel is used to prevent using
+    /// too much memory unexpectedly.
+    /// In the working thread, the `TimestampRequest`, which is actually a one channel sender,
+    /// is used to send back the timestamp result.
+    request_tx: mpsc::Sender<TimestampRequest>,
+    close_rx: watch::Receiver<()>,
+}
+
+impl TimestampOracle {
+    pub(crate) fn new(
+        cluster_id: u64,
+        pd_client: &PdClient,
+        call_option: CallOption,
+    ) -> Result<TimestampOracle> {
+        let (request_tx, request_rx) = mpsc::channel(MAX_BATCH_SIZE);
+        let (rpc_sender, rpc_receiver) = pd_client.tso_opt(call_option)?;
+        let (close_tx, close_rx) = watch::channel(());
+
+        // Start a background thread to handle TSO requests and responses
+        thread::Builder::new()
+            .name("tso-worker".into())
+            .spawn(move || {
+                block_on(run_tso(
+                    cluster_id,
+                    rpc_sender.sink_err_into(),
+                    rpc_receiver.err_into(),
+                    request_rx,
+                    close_tx,
+                ))
+            })
+            .expect("unable to create tso worker thread");
+
+        Ok(TimestampOracle {
+            request_tx,
+            close_rx,
+        })
+    }
+
+    pub(crate) fn get_timestamp(&self) -> impl Future<Output = Result<TimeStamp>> + 'static {
+        let (request, response) = oneshot::channel();
+        let request_tx = self.request_tx.clone();
+        async move {
+            request_tx
+                .send(request)
+                .await
+                .map_err(|_| -> Error { box_err!("TimestampRequest channel is closed") })?;
+            response
+                .await
+                .map_err(|_| box_err!("Timestamp channel is dropped"))
+        }
+    }
+
+    pub(crate) fn closed(&self) -> impl Future<Output = ()> {
+        let mut close_rx = self.close_rx.clone();
+        async move {
+            let _ = close_rx.changed().await;
+        }
+    }
+}
+
+async fn run_tso(
+    cluster_id: u64,
+    mut rpc_sender: impl Sink<(TsoRequest, WriteFlags), Error = Error> + Unpin,
+    mut rpc_receiver: impl Stream<Item = Result<TsoResponse>> + Unpin,
+    mut request_rx: mpsc::Receiver<TimestampRequest>,
+    close_tx: watch::Sender<()>,
+) {
+    // The `TimestampRequest`s which are waiting for the responses from the PD server
+    let pending_requests = Rc::new(RefCell::new(VecDeque::with_capacity(MAX_PENDING_COUNT)));
```

Comment:
do we have a metrics for this?

Reply:
Not yet. Nice to have, let me add it.

Reply:
Now I add a gauge and a grafana panel for it. (In normal cases, the value is always 0...)

Reply:
@Connor1996 PTAL again when you're free.